File: 0337-house-robber-iii/0337-house-robber-iii.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def rob(self, root: Optional[TreeNode]) -> int:
      # Helper function
      def accomplice(node):
        if not node:
          return [0, 0]
        
        left_side = accomplice(node.left)
        left_with, left_without = left_side

        right_side = accomplice(node.right)
        right_with, right_without = right_side
        
        with_node = node.val + left_without + right_without
        without_node = max(left_with, left_without) + max(right_with, right_without)
      
        return [with_node, without_node]
     
      # main return
      return max(accomplice(root)[0], accomplice(root)[1])

File: 0337-house-robber-iii/0337-house-robber-iii.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var rob = function(root) {
    // DFS + Memoization w/ Map (key: node, val: max $)
    const memo = new Map();
    return robMax(root);
    
    function robMax(node) {
        if (memo.has(node)) return memo.get(node);
        if (!node) return 0;
        /*
            Choose curr + :
                Decision1: curr.left.left + curr.left.right
                Decision2: curr.right.left + curr.righ.right
        */
        let chooseHouse = node.val;
        if (node.left) chooseHouse += robMax(node.left.left) + robMax(node.left.right)
        if (node.right) chooseHouse += robMax(node.right.left) + robMax(node.right.right)
        
        // Skip curr
        const skipHouse = robMax(node.left) + robMax(node.right);
        
        memo.set(node, Math.max(chooseHouse, skipHouse));
        return memo.get(node);
    }
    
};